### ISSUE TITLE
Spike: Reconsider approach to breakpoints

### ISSUE THREAD

**Issue by mcarrano** [patternfly/patternfly#5111]
Current method for defining responsive breakpoints at the viewport/page level does not always yield good results (i.e. elements being hidden at unexpected widths, etc). Breakpoints at the container level might make more sense, but this issue needs more investigation to determine if our current system of breakpoints can or should be improved.

cc @mceledonia @mcoker Anything you would add here?

**Comment by mcoker** [comment]
I think part of the discussion for this started around whether our current viewport breakpoint sizes should be updated to work better with our components overall, or if we feel like the current breakpoints work well for the most part. So it might be good to look at the current sizes and if we should change/add/remove any, if we should use different units (em vs px, for example), etc. Also take into account the `--on-[breakpoint-size]` modifiers we have and if those make sense as good options, since those are based on our global breakpoints.

We have a PR up currently to offer an alternative set of breakpoints that do not account for the sidebar/vertical nav width. These breakpoints would basically be used in a patternfly app that was full-screen and didn't have a sidebar or vertical nav. Since our current sidebar is hidden at less than our xl breakpoint of 1200px ([`$pf-global--breakpoint--xl`](https://github.com/patternfly/patternfly/blob/4d05197dab6ea210297e64b3388cb96fcf7ebc0c/src/patternfly/sass-utilities/scss-variables.scss#L167)), if you're at the lg breakpoint (992px), components have roughly 992px space to display. @ 1200px, they have roughly 800px to display (1200px - 300px for the sidebar). And @ 1450px (the 2xl) breakpoint, they have 1150px to display. How might we be able to change the breakpoint values to account for something like the absence of a sidebar and the components behave properly? Maybe we should we start with a global set of breakpoints that are optimized for full-screen as a baseline, then add support for the presence of the sidebar on top of that?

Also a really great to have would be the ability for users to customize the breakpoints without having to recompile our SCSS. If they could write their own media queries that set like `@media (min-width: [whatever value]) { --pf-global--breakpoint--sm--on: ;' }`, and we don't have a bunch of media query blocks in our individual components, the responsive CSS just references `--pf-global--breakpoint--[size]--on` to apply the correct value to another property. 

And as @mcarrano mentioned, the ability to have container based responsive design would be great. Each component could have responsive support at its own set of breakpoints - art direction at the component level basically. That's something that could potentially partially be resolved by letting users specify their own breakpoints and setting breakpoint vars to toggle the breakpoint on and off, which they could do by class or something to target individual instances of a component on a page, but that gets tricky with inherited components. Also container queries is on the horizon, with support in most browsers, which will give us the native ability to do this - https://caniuse.com/css-container-queries

**Comment by mattnolting** [comment]
### TLDR; 
Using container queries and resize observer would replace our current approach to {breakpoint} suffixed layout modifiers, resulting in significantly reduced CSS filesize/s. Each instance of breakpoint based modifiers generates excess css, which tends to be very specific (`.pf-m-spacer-lg-on-xl`) and generally goes unused. Using container queries  ([with polyfill](https://css-tricks.com/a-new-container-query-polyfill-that-just-works/)) will handle most of our basic layout shifts (think stack to inline at tablet size). Because container query values can’t use css variables (`@container wrapper (min-width: var(--pf-global--breakpoint--md))` is invalid), base modifiers could be applied via resize observer. Using resize observer to apply modifiers, at user specified container sizes, would result in simpler and more accurate application of layout modifications, independent of other, non-main content elements being visible (`.pf-c-page__sidebar`). 

**For example**: removing breakpoint suffixes in these layouts/components results in
* **Masthead**: (47k -> 25k)
* **Flex**: (45k -> 7k)
* **Toolbar**: (46k -> 20k)

### CSS
* Core CSS could be split between a Core css and optional breakpoint suffixed CSS
  * Set a SCSS `@if $pf-include-responsive-modifiers` flag that enables compilation of responsive CSS, for consumers compiling their own CSS

![Screenshot 2022-11-21 at 10 33 22 AM](https://user-images.githubusercontent.com/5385435/203102340-2fde94d3-2255-4e58-9135-c06f93c6fb4b.png)

  * Provide additional CSS for consumers using compiled CSS `.patternfly-responsive.css`

![Screenshot 2022-11-21 at 11 07 24 AM](https://user-images.githubusercontent.com/5385435/203102794-d317d928-28d8-4d19-b17c-4f228f3fca40.png)

### Container queries + polyfill 
* General layout changes, should be overwritable with a simple CSS selector  (.pf-m-inline instead of .pf-m-inline-xl-on-xl)
• Use container query polyfill, until broadly supported, for basic layout shifts (stack to inline)
• Load breakpoint suffixed CSS as an opt in, rather than by default

### Resize observer
* Could be applied to any container, where contents of the container should respond to the closest resize observer enabled on parent container. 
* By using resize observer we can add/remove a basic, non-suffixed selector (`.pf-m-spacer-lg`) based on user defined width/height container values, enabling consumers to fine-tune their applications. 

A current example of resize observer in use exists in the [card view demo](https://www.patternfly.org/v4/demos/card-view/react-demos/card-view/). 

![Screenshot 2022-11-21 at 10 53 33 AM](https://user-images.githubusercontent.com/5385435/203099587-36dc2fb7-9528-4645-82b6-e28e9234525f.png)

Resize observer applies `.pf-m-display-inline` at a user specified breakpoint, rather than default, media-query based breakpoint. 

![Screenshot 2022-11-21 at 10 51 29 AM](https://user-images.githubusercontent.com/5385435/203099629-67ee4861-69bb-4360-93d0-fc3aa2d3cd6b.png)

Caveats:
* Many components and layouts will require container wrappers. 
* Might consider creating a global container `.pf-container` that can be used to wrap any layout/component as the container query will apply styles to components based on the size of the nearest ancestor with a containment context.

**Comment by mcarrano** [comment]
@mattnolting @mcoker is this something you think we should consider doing for the v5 release or do you see this as a future initiative?

**Comment by mcoker** [comment]
I'm good with what @mattnolting wants to do, but IMO I don't think it's feasible for v5.

**Comment by garrett** [comment]
Note: Container queries will be landing in Firefox 110 (14 Feb 2023), which is the next version of Firefox. So it'll be supported in all of the big evergreen browsers within a few days.

https://developer.mozilla.org/en-US/docs/Mozilla/Firefox/Releases/110#css

Considering the timeframe (most likely to be implemented well after Valentine's Day), any fallbacks (including a polyfill) would mainly be targeting old versions of Firefox at this point, notably Firefox ESR (Extended Support Release), up until 2023-09-26. I'd suggest not using a polyfill, but making sure it doesn't completely fall apart when a browser doesn't support container queries.

In other words, either have fixed size (and wrap), use flex or grid with sizing, or use something like widths with min(), max(), or clamp() — all depending on where the container query would be used.

Since we'd only care about Firefox ESR at that point in time — and only for a limited time as [Firefox ESR will hit 115.3 on 2023-09-26](https://wiki.mozilla.org/Release_Management/Calendar) — I'd suggest keeping any fallback as simple as possible (which is probably using a standard width with wrapping, so a layout isn't the same, but also isn't broken). This would also mainly affect smaller sizes too, and Firefox ESR isn't available phones, so in most cases, it would mainly be a moot point.

Note: Firefox ESR is what versions of RHEL (and other Enterprise Linux distros) ship and what people on Windows and Mac are using in many corporate settings. So it does matter.

**Comment by garrett** [comment]
This was marked as completed... but I don't see a PR. I'd like to track the changes, as this directly affects a big issue we've been struggling with Cockpit for years. Thanks!

Where can I find out more?

**Comment by mattnolting** [comment]
Currently, the Insights team is working on a performant approach that leverages very lightweight JS and resize observer. This solution will be added to an updated proposal which will be shared with the core team, then with the community for feedback and testing. [Placeholder issue is here](https://github.com/patternfly/patternfly/issues/6660)

**Comment by garrett** [comment]
Why would you want to use some JavaScript for something that's provided natively by CSS and is in all browsers now (container queries)? Using JavaScript for something like this that CSS provides natively will always be more heavyweight and error prone.

Additionally, container queries should be really easy to port over to from the existing full page media queries. (I did it a while back on a test somewhere... so I'm speaking from experience. It'd take some time to find it again, however.)

It's early morning and I haven't had my coffee yet, but I think it was a matter of adding:

```css
.width-reference-element-here {
  container-type: size;
}
```

And then changing something like

```css
@media (max-width: 70em) {
   /* Rules here */
}
```

to

```css
@container (width >= 70em) {
   /* Rules here */
}
```

And then subtract the width of the sidebar from the magic breakpoint numbers.

And that's it!

Ideally, you'd refactor so you wouldn't need `>=` instead of `max-width` and `<=` instead of `min-width`, but this would be a drop-in replacement. (You'd want to use > and < instead of <= and >=, but this is about emulating what's already there in a first pass with minimal changes.) 

The object that would be the reference would be the main page element that contains the components. (So the content is the target, not the 

I thought the only thing that was blocking going to container queries was waiting for full browser support, which happened last year (summer 2022 for Chrome and Safari, Feb 14 2023 for Firefox 110), and then waiting [a little longer for Firefox ESR](https://whattrainisitnow.com/calendar/) (2023-09-26, when 115.3 fully replaced 102.15) support.

So we've been in the clear to use container size queries for almost a year even on older enterprise browsers that we're shipping to customers in RHEL. (And more than 2 years already in Chrome and WebKit, more than 1 year in Firefox.)

Also, what this is closed with a replacement for... @ https://github.com/patternfly/patternfly/issues/6660 I don't see any code or reasoning why JavaScript would be preferred. It's really surprising and a bit confusing.

Reopening, as:

1. Container size queries really are the obvious choice.
2. Container size queries have been supported everywhere for quite some time.
3. JavaScript doesn't belong in this area.
4. There's no justification why container queries are being replaced with fragile JavaScript.
5. There's no code example as to how the JavaScript listener would work.
6. This is something that Cockpit needs from PatternFly and I've been invested in this discussion for a few years already, just to have this abruptly closed without an explanation as to why or how.

Please explain the above if you'd like to re-close it. Or, perhaps we can talk about a standards-based approach instead of throwing yet more JavaScript at a problem that's already been solved natively by browsers for a while now.

**Comment by garrett** [comment]
(Also, general "rule of thumb": Issues really shouldn't be closed until there's code merged into the codebase that actually solves the problem.)

**Comment by mattnolting** [comment]
> Why would you want to use some JavaScript for something that's provided natively by CSS and is in all browsers now (container queries)? Using JavaScript for something like this that CSS provides natively will always be more heavyweight and error prone.

@garrett several reasons
- Container queries require a superfluous wrapping element. Anything placed specifically in a page layout is now broken. 
- In the context of an enterprise applications, this construct would add a tremendous amount of DOM bloat. 
- Container queries are immutable and therefore can't be updated to suite varied use cases. 
- Can't use `absolute` positioning on child elements
- Specificity issues with adding container classes
- Inheriting from parent elements is not possible unless you use floats.
- Browsers continuously monitor container queries, even more performance impact.

> And then subtract the width of the sidebar from the magic breakpoint numbers.

Using magic numbers has more of a performance impact than this solution

> And that's it! (Yes, the solution would be to literally add a line of CSS to the parent element, then use a regular expression to change the media queries, and subtract the sidebar width from the breakpoint numbers, and it should be a drop-in replacement.)

That's not correct. This issue requires a new parent element that only describes an area in the DOM and is not otherwise stylable. 

I'm going to reopen this issue to enable implementation testing. @garrett you're correct, this should absolutely replace `media-queries`. However, container queries have implications far beyond simply adding a line of CSS.

**Comment by garrett** [comment]
Here's a quick codepen that demos media queries as I mentioned above (and we don't even change the min-width / max-width already in place in PatternFly):

https://codepen.io/garrett/pen/jOobaaq

Additionally, you can see:

1. The container is styled
2. The second child has position relative, shifting it down by 10px
3. The width is coming from the .container element, not the page (resize the page; the query is 2x the width of the children, so they'll change color if they're side by side)
4. The height query, if it existed, would come from the body (but it's not in use in this demo)

Be sure to resize the demo page!

---

Cockpit already monkeys around to rewrite the compiled CSS to change PatternFly's existing magic numbers for >= desktop sizes. (Mobile sizes do not have visible sidebars.)

https://github.com/cockpit-project/cockpit/blob/main/pkg/lib/esbuild-common.js#L8-L20

> ```js
>             // Do not override the sm breakpoint as for width < 768px the left nav is hidden
>             '768px': '428px',
>             '992px': '652px',
>             '1200px': '876px',
>             '1450px': '1100px',
> ```

---

So, what would need to be done is:

1. Add this bit of CSS (more or less):

```css
body {
    /* Used for vertical page sizes, or we can exclude this and use @media still for vertical and viewports only */
    container-type: size;
}

.pf-v5-c-page {
    /* Set up width container for @container size queries */
    container-type: inline-size;
}
```

2. Then switch `@media` to `@container`. Keep the part inside (), such as `min-width`, `max-width` the same.

3. Change the number definitions in the SCSS, so 768 becomes 428, and so on (see above).

And that should be it.

---

As my demo shows, none of those drawbacks exist within the scope of using container size queries instead of media queries for breakpoint layouts:


>    Container queries require a superfluous wrapping element. Anything placed specifically in a page layout is now broken.

They don't require an extra wrapping element.

>    In the context of an enterprise applications, this construct would add a tremendous amount of DOM bloat.

PatternFly has a tremendous amount of DOM bloat as is. This isn't an argument at all, especially since what I'm talking about doesn't affect the DOM at all.

>    Container queries are immutable and therefore can't be updated to suite varied use cases.

I'm only talking about them in the context of switching away from media queries in breakpoints to retarget the current breakpoints to be tied to the content area instead of the viewport.

We could also consider container queries for other things, or observers, or flex, or grid, or whatever else.  

>  Can't use absolute positioning on child elements

We can, actually! My demo uses relative and absolute positioning on the direct children.

>    Specificity issues with adding container classes

No more than media queries, which I'm suggesting we replace with container size queries.

>    Inheriting from parent elements is not possible unless you use floats.

Nope, my demo inherits a CSS custom property from the .container and it's applied in the child in the container size query.

>    Browsers continuously monitor container queries, even more performance impact.

I'd like to see a reference to this. 

I doubt it's any more than with normal media queries. 

If you're talking about replacing it with a _JavaScript_ listener — well that will always have a detrimental performance impact, and it doesn't scale well. (JavaScript is not suited for things CSS natively handles.) JavaScript listeners are _notorious_ for performance issues (especially depending on what they do).

---

To be absolutely clear:

I'm not saying we should use container size queries everywhere, nor am I saying that we shouldn't use resize observers. This is also not about layouts that might be better suited for flex or grid. It's also not about using other methods for sizing, like `min()`, `max()`, `clamp()`, etc.

This is specifically about **changing media queries to container size queries so that we don't have sizing issues when sidebars aren't used**. (I _think_ this also affects wizards, drawers, and other things outside the content area.)

We can — _and should_ — continue the discussions about other techniques for sizing outside of this context, but I think it's pretty clear that it really looks to be be a big win without drawbacks and without much effort to be able to finally fix the sidebar vs. no sidebar breakpoint issue. (Media size queries might be an answer sometimes, resize observers might be another answer, and we might want to use grid or min, max, clamp, etc. We'd need to have _those_ conversations on a case-by-case basis.)

**Comment by Venefilyn** [comment]
Just posting a comment here for clarity's sake so we don't forget it. Me and Garrett had a meeting with some PF team members and did bring up this as an important thing for us that we'd ideally want this year at least.

For Cockpit it would help us with our iframes and respecting window width better.

Our usage in Cockpit causes current default breakpoints to not work when you have content on a webpage and some content inside an iframe. Due to the media rules currently in place they trigger the media queries at different stages and are not synchronized.
